src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -30,179 +30,171 @@ protected MessageBody(Frame context)
 
         public bool RequestUpgrade { get; protected set; }
 
-        public Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken = default(CancellationToken))
+        public virtual bool IsEmpty => false;
+
+        public virtual async Task StartAsync()
         {
-            var task = PeekAsync(cancellationToken);
+            Exception error = null;
 
-            if (!task.IsCompleted)
+            try
             {
-                TryProduceContinue();
+                while (true)
+                {
+                    var awaitable = _context.Input.ReadAsync();
 
-                // Incomplete Task await result
-                return ReadAsyncAwaited(task, buffer);
-            }
-            else
-            {
-                var readSegment = task.Result;
-                var consumed = CopyReadSegment(readSegment, buffer);
+                    if (!awaitable.IsCompleted)
+                    {
+                        TryProduceContinue();
+                    }
 
-                return consumed == 0 ? TaskCache<int>.DefaultCompletedTask : Task.FromResult(consumed);
-            }
-        }
+                    var result = await awaitable;
+                    var readableBuffer = result.Buffer;
+                    var consumed = readableBuffer.Start;
+                    // TODO: set to readableBuffer.Start after https://github.com/dotnet/corefxlab/issues/1540 is fixed
+                    var examined = readableBuffer.IsEmpty ? readableBuffer.End : readableBuffer.Start;
 
-        private async Task<int> ReadAsyncAwaited(ValueTask<ArraySegment<byte>> currentTask, ArraySegment<byte> buffer)
-        {
-            return CopyReadSegment(await currentTask, buffer);
-        }
+                    try
+                    {
+                        if (!readableBuffer.IsEmpty)
+                        {
+                            var writableBuffer = _context.RequestBodyPipe.Writer.Alloc(1);
+                            bool done;
 
-        private int CopyReadSegment(ArraySegment<byte> readSegment, ArraySegment<byte> buffer)
-        {
-            var consumed = Math.Min(readSegment.Count, buffer.Count);
+                            try
+                            {
+                                done = Read(readableBuffer, writableBuffer, out consumed, out examined);
+                            }
+                            finally
+                            {
+                                writableBuffer.Commit();
+                            }
+
+                            await writableBuffer.FlushAsync();
 
-            if (consumed != 0)
+                            if (done)
+                            {
+                                break;
+                            }
+                        }
+                        else if (result.IsCancelled)
+                        {
+                            break;
```

Comment:
Frame.StopAsync() calls Input.CancelPendingRead(), but we don't want that to kill the request body. We only want to do that given an abort.